apps/prairielearn/src/pages/enroll/enroll.ts

```diff
@@ -66,29 +70,38 @@ router.post(
         // Enrollment limits can only be configured on enterprise instances, so
         // we'll also only check and enforce the limits on enterprise instances.
         if (isEnterprise()) {
-          const {
-            institution,
-            course_instance,
-            institution_enrollment_count,
-            course_instance_enrollment_count,
-          } = await queryRow(
-            sql.select_and_lock_enrollment_counts,
+          const { course, course_instance } = await queryRow(
+            sql.select_course_instance,
             { course_instance_id: req.body.course_instance_id },
             z.object({
-              institution: InstitutionSchema,
+              course: CourseSchema,
               course_instance: CourseInstanceSchema,
-              institution_enrollment_count: z.number(),
-              course_instance_enrollment_count: z.number(),
             }),
           );
+          const institution = await queryRow(
+            sql.select_and_lock_institution,
+            { institution_id: course.institution_id },
+            InstitutionSchema,
+          );
+
+          const institutionEnrollmentCounts = await getEnrollmentCountsForInstitution({
+            institution_id: institution.id,
+            created_since: '1 year',
+          });
+          const courseInstanceEnrollmentCounts = await getEnrollmentCountsForCourseInstance(
+            req.body.course_instance_id,
+          );
+
+          const freeInstitutionEnrollmentCount = institutionEnrollmentCounts.free;
+          const freeCourseInstanceEnrollmentCount = courseInstanceEnrollmentCounts.free;
 
           const yearlyEnrollmentLimit = institution.yearly_enrollment_limit;
           const courseInstanceEnrollmentLimit =
             course_instance.enrollment_limit ?? institution.course_instance_enrollment_limit;
 
           if (
-            institution_enrollment_count + 1 > yearlyEnrollmentLimit ||
-            course_instance_enrollment_count + 1 > courseInstanceEnrollmentLimit
+            freeInstitutionEnrollmentCount + 1 > yearlyEnrollmentLimit ||
+            freeCourseInstanceEnrollmentCount + 1 > courseInstanceEnrollmentLimit
```

Comment:
We'll have to be careful before we actually ship the student payment page. As currently implemented, if an institution or course instance is _at_ their enrollment limit, we'll be unable to enroll the student even if that student won't actually count towards the limit (because they'll be paying).

Probably the right way to handle this is to not create the enrollment until when we actually create the plan grant. We'll need some kind of pre-enrollment flow in that case.

As I think this through, I realize there's a big hole in our current approach: from looking at our authz code, I think students will be automatically enrolled when they land on a course instance page, even if they've never visited the enrollment page. We'll want to revisit that so that we can enforce enrollment limits correctly (and so that we have a place to bounce students to a payment page). I'm imagining a route like `/pl/enroll/:course_instance_id` that we can bounce students to from our authz middleware. If no payment or anything special is required, this can have a straightforward "Enroll" button. Otherwise, this is where we can show a payment button, and ensure that we enforce any enrollment limits.

Reply:
Good point, at the moment we do auto enrollment on deep linking but I guess this will have to stop. Are you thinking that we'll bounce any unenrolled students to the enrollment page with a query parameter for where to redirect them afterwards? If payment is required then we can abandon this query parameter, while using it for courses that don't require payment (like some of our example/template courses).

Reply:
Maybe as a tradeoff/stopgap, we can continue auto-enrolling students if they can do so without taking additional steps (read: paying), but otherwise bounce them to this new enrollment page?

Reply:
https://github.com/PrairieLearn/PrairieLearn/pull/8269 laid the groundwork for our desired auto-enrollment scheme. I'm not going to address that in this PR, but we should be able to fast-follow this by lifting the limit-aware enrollment code out into a reusable function and calling it from `authzCourseOrInstance` (or redirecting appropriately).

Reply:
Sounds good to me!